### PR TITLE
(Do not merge) Fhir v1.4

### PR DESCRIPTION
**Pull request by luislee818**
To show the changes for supporting FHIR v1.4.

### CODE REVIEW

**Comment by radamson** [comment]
A FHIR R4 (version 4.0.1) version of this library already exists: https://github.com/fhir-crucible/fhir_models

This library is for FHIR STU3 (version 3.0.0).

**Comment by luislee818** [comment]
@radamson sorry, I selected the wrong base branch, I meant to select `master` of my fork as base.

Yes, I understand this repo is for STU3. I was trying to create a snapshot that works with the [STU3 v1.4.0](http://hl7.org/fhir/2016May/) version, which the vendor uses I'm working on an integration.

I will close this, sorry about the confusions.

Just in case you have any suggestions, do you think my approach (generating the resource classes from the schema definitions of v1.4.0) is the way to go? Thanks!

**Comment by radamson** [comment]
@luislee818 no problem, just throwing it out there in case that was what you were looking for!  STU3 v1.4.0 is a pretty unique version.

This library, and its sister libraries, build the resource classes off the StructureDefinitions in the spec.  I haven't looked at your code specifically (the diff is too large to see in GitHub at a glance 😄 ), but at a high level I think its the way to go and what many FHIR libraries in other languages do.  You can get a similar result working off the schema vs the StructureDefinition, but the StructureDefinition's have more metadata that is useful for other activities, like validation.

Happy to discuss further and good luck with implementing v1.4.0!

**Comment by luislee818** [comment]
@radamson thanks for the explanation! While I have your attention, hope you don't mind I ask a couple of more questions, thanks in advance!

#### regarding getting the repo to support an old version of FHIR
Yeah, the PR is too large to be conveniently reviewed on Github (or anywhere...). What I did to support an old version of FHIR using this repo was:
1. find the the schema definitions in an old commit matching the version I would like to support (v1.4.0 in this case, from [a commit back in 2016](https://github.com/fhir-crucible/fhir_stu3_models/tree/ba058c4f8c3df166974ccca9e4d522051d1f2f38))
2. checkout the latest version of the repo, overwrite the schema definitions (everything inside the `definitions` folder)
3. run the rake task to generate resource classes (`bundle exec rake fhir:generate`)
4. also overwrite the files in `examples` from the old commit

I probably don't fully understand the desired workflow or the logic of the repo, but those are the steps I could come up with. Maybe in step no.1 I could also obtain the schema definitions of a specific version by using the [FHIR build tools](https://github.com/HL7/fhir) and some rake tasks in the repo, not sure if that's more appropriate. 🤷‍♂️ 

#### regarding an alternative approach
You mentioned another approach could be
> ...can get a similar result working off the schema...

Do you mean updating the resource classes under `fhir/resources`, like [`Practitioner.rb`](https://github.com/fhir-crucible/fhir_stu3_models/blob/master/lib/fhir_stu3_models/fhir/resources/Practitioner.rb) directly? I was initially thinking about doing that as that's straightforward and simple, though as one of my colleagues suggested that's also kind of "hacky", no? As to get a faithful v1.4.0 version of the resources, one would need to manually update all the resources, right? 🙂

**Comment by radamson** [comment]
That workflow looks right to me.  Just to confirm you are updating everything in the `/definitions` directory correct?  There are a number of sub-directories and the key one for generating the new models is `/definitions/structures`.  Getting them from the build tools would also be appropriate as that's the more canonical source, but I wouldn't expect a difference.

**In terms of the "alternate approach"...**

I thought you were talking about the FHIR XML schema's when you said:

> generating the resource classes from the schema definitions

In `/definitions` you'll notice two key subdirectories: `/definitions/schema` and `/definitions/structures`

The artifacts in `/definitions/structures` are used in the code/class generation.  The artifacts in `/definitions/schema` are used for validation of XML resources.  Apologies for the confusion there. I definitely wouldn't suggest updating all the resources manually either.

----
TL;DR: Your approach looks good just make sure you are updating all the artifacts in the `definitions` directory, including the subdirectories.

**Comment by luislee818** [comment]
@radamson yes, I was copying everything inside the `/definitions` directory.

My apologies for not being clear, I'm still new to FHIR and would often use inaccurate terms.

Thanks again for your answers! 🙏
